pages/userSettings/userSettings.js

```diff
@@ -0,0 +1,72 @@
+const ERR = require('async-stacktrace');
+const express = require('express');
+const router = express.Router();
+const crypto = require('crypto');
+const uuidv4 = require('uuid/v4');
+
+const error = require('@prairielearn/prairielib/error');
+const sqldb = require('@prairielearn/prairielib/sql-db');
+const sqlLoader = require('@prairielearn/prairielib/sql-loader');
+
+const sql = sqlLoader.loadSqlEquiv(__filename);
+
+router.post('/', (req, res, next) => {
+    if (req.body.__action === 'token_generate') {
+        const name = req.body.token_name;
+        const token = uuidv4();
+        const tokenHash = crypto.createHash('sha256').update(token, 'utf8').digest('hex');
+
+        const params = {
+            user_id: res.locals.authn_user.user_id,
+            name,
+            // The token will only be persisted until the next page render
+            // After that, we'll remove it from the database
+            token,
+            token_hash: tokenHash,
+        };
+        sqldb.queryOneRow(sql.insert_access_token, params, (err) => {
+            if (ERR(err, next)) return;
+            res.redirect(req.originalUrl);
```

Comment:
Rather than doing a redirect, we could actually just render the page directly here. That way we wouldn't need to do this dance of store-token-then-delete. I normally prefer the POST/redirect/GET pattern (see https://prairielearn.readthedocs.io/en/latest/dev-guide/#state-modifying-post-requests and https://en.wikipedia.org/wiki/Post/Redirect/Get) but this is special case where rendering the page in the POST handler might actually be preferable?

Reply:
If we don't use PRG here, refreshing after creating a token would cause another token to be created in most user agents. I would definitely like to not be doing this little dance, but I do think we still need PRG here.

Reply:
Agreed.